### PR TITLE
armv7-a/r:cache: add cp15_cache_size function and use flush/clean_all if size large than cache size

### PR DESCRIPTION
**Pull request by zhuyanlinzyl**
patch 1 : Add cp15_cache_size function for armv7-a/r
patch 2: use flush/clean_all if size large than cache size

## Summary
For better cache performance. when flush/clean cache size large than cache size, use flush/clean_all instead 

## Testing

Testing in Qemu , cache size  16K
Testing in our smp armv7-a board, cache size 32K

### CODE REVIEW

File: arch/arm/src/armv7-a/cp15_cache_size.S [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
```
Comment (acassis) [inline arch/arm/src/armv7-a/cp15_cache_size.S:40]: Why are you using this license? If you are creating a file from scratch the license should be Apache

File: arch/arm/src/armv7-r/cp15_cache_size.S [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
```
Comment (acassis) [inline arch/arm/src/armv7-r/cp15_cache_size.S:39]: ditto

Comment (zhuyanlinzyl) [reply arch/arm/src/armv7-r/cp15_cache_size.S]: @acassis  I update the patch, use Apache license instead

File: arch/arm/src/armv7-a/arm_cache.c [hunk 3]
```diff
@@ -134,8 +134,16 @@ void up_invalidate_icache_all(void)
 
 void up_clean_dcache(uintptr_t start, uintptr_t end)
 {
-  cp15_clean_dcache(start, end);
-  l2cc_clean(start, end);
+  if (cp15_cache_size() < (end - end))
+    {
+      cp15_clean_dcache(start, end);
+      l2cc_clean(start, end);
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-a/arm_cache.c:140]: l2cc_clean need check by l2cc_cache_size api

File: arch/arm/src/armv7-a/arm_cache.c [hunk 4]
```diff
@@ -190,8 +198,16 @@ void up_clean_dcache_all(void)
 
 void up_flush_dcache(uintptr_t start, uintptr_t end)
 {
-  cp15_flush_dcache(start, end);
-  l2cc_flush(start, end);
+  if (cp15_cache_size() < (end - end))
+    {
+      cp15_flush_dcache(start, end);
+      l2cc_flush(start, end);
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-a/arm_cache.c:204]: ditto

File: arch/arm/src/armv7-a/arm_cache.c [hunk 5]
```diff
@@ -190,8 +198,16 @@ void up_clean_dcache_all(void)
 
 void up_flush_dcache(uintptr_t start, uintptr_t end)
 {
-  cp15_flush_dcache(start, end);
-  l2cc_flush(start, end);
+  if (cp15_cache_size() < (end - end))
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-a/arm_cache.c:201]:
```suggestion
  if (cp15_cache_size() < (end - start))
```

Comment (pkarashchenko) [reply arch/arm/src/armv7-a/arm_cache.c:201]: I think that test cases have a big gap if this was not catched during testing.

File: arch/arm/src/armv7-r/arm_cache.c [hunk 6]
```diff
@@ -134,8 +134,16 @@ void up_invalidate_icache_all(void)
 
 void up_clean_dcache(uintptr_t start, uintptr_t end)
 {
-  cp15_clean_dcache(start, end);
-  l2cc_clean(start, end);
+  if (cp15_cache_size() < (end - end))
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-r/arm_cache.c:137]:
```suggestion
  if (cp15_cache_size() < (end - start))
```

File: arch/arm/src/armv7-r/arm_cache.c [hunk 7]
```diff
@@ -190,8 +198,16 @@ void up_clean_dcache_all(void)
 
 void up_flush_dcache(uintptr_t start, uintptr_t end)
 {
-  cp15_flush_dcache(start, end);
-  l2cc_flush(start, end);
+  if (cp15_cache_size() < (end - end))
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-r/arm_cache.c:201]:
```suggestion
  if (cp15_cache_size() < (end - start)
```

File: arch/arm/src/armv7-a/arm_cache.c (hunk 5 above)
Comment (zhuyanlinzyl) [reply arch/arm/src/armv7-a/arm_cache.c]: I think if the buffer being flushed (end - start)  larger than cache size,  the entire cache is used by the buffer.

File: arch/arm/src/armv7-r/arm_cache.c [hunk 8]
```diff
@@ -134,7 +134,15 @@ void up_invalidate_icache_all(void)
...
+    {
+      cp15_clean_dcache(start, end);
+    }
+  else
+    {
+      cp15_clean_dcache_all();
+    }
+
   l2cc_clean(start, end);
 }
```
Comment (xiaoxiang781216) [inline arch/arm/src/armv7-r/arm_cache.c:147]: let's optimize l2cc cache too

Comment (zhuyanlinzyl) [reply arch/arm/src/armv7-r/arm_cache.c:147]: l2cc have already implemention in l2cc_flush l2cc_clean function

Comment (xiaoxiang781216) [reply arch/arm/src/armv7-r/arm_cache.c:147]: Nice!

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/arm/src/armv7-a/arm_cache.c [hunk 9]
```diff
@@ -190,7 +198,15 @@ void up_clean_dcache_all(void)
 
 void up_flush_dcache(uintptr_t start, uintptr_t end)
 {
-  cp15_flush_dcache(start, end);
+  if (cp15_cache_size() < (end - start))
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-a/arm_cache.c:201]: I have a question about how optimal is current implementation if memory from `start` till `end` is not currently in cache? Or some part is in cache and other part is not?

Comment (zhuyanlinzyl) [reply arch/arm/src/armv7-a/arm_cache.c:201]: We have a testing on armv8-m with 32K icache & 32k dcache, when flush &clean buffer with size 1MB。

up_clean_dcache_all() run 30 times faster than up_clean_dcache().

When use this patch, the time is same.

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**